algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala

```diff
@@ -441,23 +441,90 @@ case class CMSZero[K](override val params: CMSParams[K]) extends CMS[K](params)
  */
 case class CMSItem[K](item: K, override val totalCount: Long, override val params: CMSParams[K]) extends CMS[K](params) {
 
-  override def +(x: K, count: Long): CMS[K] = CMSInstance[K](params) + item + (x, count)
+  override def +(x: K, count: Long): CMS[K] = SparseCMS[K](params) + (item, totalCount) + (x, count)
 
   override def ++(other: CMS[K]): CMS[K] = {
     other match {
       case other: CMSZero[_] => this
-      case other: CMSItem[K] => CMSInstance[K](params) + item + other.item
-      case other: CMSInstance[K] => other + item
+      case other: CMSItem[K] => CMSInstance[K](params) + (item, totalCount) + (other.item, other.totalCount)
+      case _ => other + item
     }
   }
 
-  override def frequency(x: K): Approximate[Long] = if (item == x) Approximate.exact(1L) else Approximate.exact(0L)
+  override def frequency(x: K): Approximate[Long] = if (item == x) Approximate.exact(totalCount) else Approximate.exact(0L)
 
-  override def innerProduct(other: CMS[K]): Approximate[Long] = other.frequency(item)
+  override def innerProduct(other: CMS[K]): Approximate[Long] = Approximate.exact(totalCount) * other.frequency(item)
 
 }
 
 /**
+ * A sparse Count-Min sketch structure, used for situations where the key is highly skewed.
+ */
+case class SparseCMS[K](exactCountTable: Map[K, Long], maxExactCountOpt: Option[Int] = None,
```

Comment:
Make sense. Has made the change as suggested.

On Tue, Jul 28, 2015 at 10:41 AM, P. Oscar Boykin <notifications@github.com>
wrote:

> In algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala
> <https://github.com/twitter/algebird/pull/464#discussion_r35676556>:
>
> >
> >  }
> >
> >  /**
> > + * A sparse Count-Min sketch structure, used for situations where the key is highly skewed.
> > + */
> > +case class SparseCMS[K](exactCountTable: Map[K, Long], maxExactCountOpt: Option[Int] = None,
>
> I agree it is only relevant to one part of the state machine, but
> similarly, the size is only depth/width is only relevant to when we create
> the hash count matrix (not to CMSItem or SparseCMS). So, I think it still
> fits as a parameter of the operation of the monoid.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/algebird/pull/464/files#r35676556>.
>



-- 
Kevin Lin | Twitter, Inc.
1355 Market St. | San Francisco, CA | 94103

Follow me: @reconditesea <https://twitter.com/reconditesea>